nautobot_secrets_providers/providers/thycotic.py

```diff
@@ -0,0 +1,247 @@
+        except KeyError as err:
+            msg = f"The secret parameter could not be retrieved for field {err}"
+            raise exceptions.SecretParametersError(secret, cls, msg) from err
+
+        if secret_id is None and secret_path is None:
+            msg = "The secret parameter could not be retrieved for field!"
+            raise exceptions.SecretParametersError(secret, cls, msg)
+
+        thycotic_settings = plugin_settings["thycotic"]
+        is_valid_base_url = "base_url" in thycotic_settings and thycotic_settings["base_url"] != ""
+        if not is_valid_base_url:
+            raise exceptions.SecretProviderError(secret, cls, "Thycotic Secret Server is not configured!")
+        is_valid_domain = "domain" in thycotic_settings and thycotic_settings["domain"] != ""
+        is_valid_password = "password" in thycotic_settings and thycotic_settings["password"] != ""
+        is_valid_tenant = "tenant" in thycotic_settings and thycotic_settings["tenant"] != ""
+        is_valid_token = "token" in thycotic_settings and thycotic_settings["token"] != ""
+        is_valid_username = "username" in thycotic_settings and thycotic_settings["username"] != ""
+        is_valid_ca_bundle_path = "ca_bundle_path" in thycotic_settings and thycotic_settings["ca_bundle_path"] != ""
+
+        return cls.query_thycotic_secret_server(
+            secret=secret,
+            base_url=thycotic_settings["base_url"],
+            ca_bundle_path=thycotic_settings["ca_bundle_path"] if is_valid_ca_bundle_path else None,
+            cloud_based=thycotic_settings["cloud_based"],
+            domain=thycotic_settings["domain"] if is_valid_domain else None,
+            password=thycotic_settings["password"] if is_valid_password else None,
+            secret_id=secret_id,
+            secret_path=secret_path,
+            secret_selected_value=secret_selected_value,
+            tenant=thycotic_settings["tenant"] if is_valid_tenant else None,
+            token=thycotic_settings["token"] if is_valid_token else None,
+            username=thycotic_settings["username"] if is_valid_username else None,
+            caller_class=cls,
+        )
+
+    @staticmethod
+    def query_thycotic_secret_server(  # pylint: disable=too-many-boolean-expressions,too-many-locals,too-many-branches,too-many-arguments
+        secret,
+        base_url,
+        ca_bundle_path=None,
+        cloud_based=None,
+        domain=None,
+        password=None,
+        secret_id=None,
+        secret_path=None,
+        secret_selected_value=None,
+        tenant=None,
+        token=None,
+        username=None,
+        caller_class=None,
+    ):
+        """Query Thycotic Secret Server."""
+        # Ensure required parameters are set
+        if (
+            token is None and (username is None or password is None)
+        ) or (  # pylint: disable=too-many-boolean-expressions
+            cloud_based and (tenant is None or username is None or password is None)
+        ):
```

Comment:
This is really hard to follow and either needs to be simplified (see my suggestion below) or broken into two separate conditions one each for `token` and one for `cloud_based` conditions.

Based on how I'm reading the code, here is my suggested readability improvement. Altho since I haven't tested it, you would need to double check that it satisfies the same intent:
```suggestion
        if any([
            token is None and not any([username, password]),
            cloud_based and not any([tenant, username, password])
        ]):
```

Reply:
This will replace the abve if

```
  if any([
    token is None and not all([username, password]),
    cloud_based and not all([tenant, username, password])
  ]):
```